### ISSUE TITLE
Segmentation fault in hash lookup in regex substitution

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9935]
Migrated from [rt.perl.org#70091](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=70091) (status was 'resolved')

Searchable as RT70091$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



\#\!perl
use warnings;                                                                        
my %foo;                                                                             
my $foo = '';                                                                        
\(my $tmp = $foo\) =~ s/^/$foo \{$0\}/e;
\_\_END\_\_

This program causes a segfault with 5\.10\.0 and 5\.10\.1\.

The space between '$foo' and '\{' is essential\, which is why piping
it through perl \-MO=Deparse "fixes" it\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.8 - Wed Mar 22 20:53:12 CET 2006
It is being executed now by  Perl v5.10.1 - Sun Sep 13 11:42:53 CEST 2009.

Site configuration information for perl v5.10.1:

Configured by mauke at Sun Sep 13 11:42:53 CEST 2009.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=2.6.27-gentoo-r8, archname=i686-linux
    uname='linux nora 2.6.27-gentoo-r8 #4 preempt mon apr 13 21:59:34 cest 2009 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.9.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.9'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    SAFEARGV0 - disable magic open in <ARGV>
    METHQUAL0 - allow qualified method calls without a valid object
    UNWARN0 - disable stupid warnings for print and qw


@INC for perl v5.10.1:
    /home/mauke/usr/local/lib/perl5/5.10.1/i686-linux
    /home/mauke/usr/local/lib/perl5/5.10.1
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.1/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.1
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.0/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.0
    /home/mauke/usr/local/lib/perl5/site_perl
    .


Environment for perl v5.10.1:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.3.2:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/opt/dmd/bin:/usr/kde/3.5/bin:/usr/qt/3/bin:/usr/games/bin:/opt/vmware/server/bin
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
@xdg - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @xdg

Confirmed that this bug does not apply to 5\.8\.9 but does apply to
5\.10\.0\, 5\.10\.1\, 5\.11\.0 and 5\.11\.1\.

**Comment by p5pRT** [comment]
### From @obra

Fixed with a series of commits culminating in 
bc795616b74a72425cddf467dbdc6e65bc76484b

**Comment by p5pRT** [comment]
@obra - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @obra

Thanks\. Applied\. Though based on discussions with a couple other
porters\, I opted not to create a new toplevel test directory\, so the
tests got moved around\. Also\, for the future\, it'd be great if tests
that test specific bugs include the bug \# in the test file or message\.

\.\.\.and actually\, it's worth mentioning that if you use \[perl \#xxxxx\]
syntax in your subject line\, your message gets automatically appended to
the bug\, which is useful\.

Best\,
Jesse


On Tue 24\.Nov'09 at 12&#8203;:50&#8203;:16 \+0100\, Gerard Goossen wrote&#8203;:
> Attached are three patches&#8203;:
> \- Adding an assertion to padav and padhv to make sure the lexical variable
>   is of the correct type\, making the bug more obvious\.
> \- The actual fix\, whitespace is also ignored when the are brackets even
>   when interpolation\.
> \- Adding a simplified version of the bug\. Because I couldn't find a good
>   place to put it\, I created a new test in "t/lex/interpolate\.t"
> 
> Gerard Goossen

> From 4cebc69f48b708aac8faaa2771c037d4b7cfbfd8 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
> Date&#8203;: Tue\, 24 Nov 2009 11&#8203;:27&#8203;:04 \+0100
> Subject&#8203;: \[PATCH 1/3\] Add assertions that pp\_padav and pp\_padhv push scalars of the correct type\.
> 
> \-\-\-
>  pp\.c |    2 \+\+
>  1 files changed\, 2 insertions\(\+\)\, 0 deletions\(\-\)
> 
> diff \-\-git a/pp\.c b/pp\.c
> index b271e7b\.\.292feb6 100644
> \-\-\- a/pp\.c
> \+\+\+ b/pp\.c
> @&#8203;@&#8203; \-63\,6 \+63\,7 @&#8203;@&#8203; PP\(pp\_padav\)
>  \{
>      dVAR; dSP; dTARGET;
>      I32 gimme;
> \+    assert\(SvTYPE\(TARG\) == SVt\_PVAV\);
>      if \(PL\_op\->op\_private & OPpLVAL\_INTRO\)
>      if \(\!\(PL\_op\->op\_private & OPpPAD\_STATE\)\)
>          SAVECLEARSV\(PAD\_SVl\(PL\_op\->op\_targ\)\);
> @&#8203;@&#8203; \-106\,6 \+107\,7 @&#8203;@&#8203; PP\(pp\_padhv\)
>      dVAR; dSP; dTARGET;
>      I32 gimme;
>  
> \+    assert\(SvTYPE\(TARG\) == SVt\_PVHV\);
>      XPUSHs\(TARG\);
>      if \(PL\_op\->op\_private & OPpLVAL\_INTRO\)
>      if \(\!\(PL\_op\->op\_private & OPpPAD\_STATE\)\)
> \-\- 
> 1\.6\.5
> 

> From eb872e07a2a07075efbfaa3d8a2776c0aa762272 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
> Date&#8203;: Tue\, 24 Nov 2009 12&#8203;:06&#8203;:55 \+0100
> Subject&#8203;: \[PATCH 2/3\] Also skip spaces after variable if we are within lexical brackets\.
>  Fixes \#70091&#8203;: Segmentation fault in hash lookup in regex substitution
> 
> \-\-\-
>  toke\.c |    2 \+\-
>  1 files changed\, 1 insertions\(\+\)\, 1 deletions\(\-\)
> 
> diff \-\-git a/toke\.c b/toke\.c
> index a4e9471\.\.f214ddf 100644
> \-\-\- a/toke\.c
> \+\+\+ b/toke\.c
> @&#8203;@&#8203; \-5441\,7 \+5441\,7 @&#8203;@&#8203; Perl\_yylex\(pTHX\)
>      d = s;
>      \{
>          const char tmp = \*s;
> \-        if \(PL\_lex\_state == LEX\_NORMAL\)
> \+        if \(PL\_lex\_state == LEX\_NORMAL || PL\_lex\_brackets\)
>          s = SKIPSPACE1\(s\);
>  
>          if \(\(PL\_expect \!= XREF || PL\_oldoldbufptr == PL\_last\_lop\)
> \-\- 
> 1\.6\.5
> 

> From d0d0240615a8364b270972683f449d8504c10700 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
> Date&#8203;: Tue\, 24 Nov 2009 12&#8203;:13&#8203;:13 \+0100
> Subject&#8203;: \[PATCH 3/3\] add interpolation test
> 
> \-\-\-
>  MANIFEST            |    1 \+
>  t/TEST              |    2 \+\-
>  t/lex/interpolate\.t |   15 \+\+\+\+\+\+\+\+\+\+\+\+\+\+\+
>  3 files changed\, 17 insertions\(\+\)\, 1 deletions\(\-\)
>  create mode 100644 t/lex/interpolate\.t
> 
> diff \-\-git a/MANIFEST b/MANIFEST
> index bc4460d\.\.7300aef 100644
> \-\-\- a/MANIFEST
> \+\+\+ b/MANIFEST
> @&#8203;@&#8203; \-4234\,6 \+4234\,7 @&#8203;@&#8203; t/io/tell\.t            See if file seeking works
>  t/io/through\.t            See if pipe passes data intact
>  t/io/utf8\.t            See if file seeking works
>  t/japh/abigail\.t        Obscure tests
> \+t/lex/interpolate\.t            See if interpolating strings work
>  t/lib/1\_compile\.t        See if the various libraries and extensions compile
>  t/lib/Cname\.pm            Test charnames in regexes \(op/pat\.t\)
>  t/lib/common\.pl            Helper for lib/\{warnings\,feature\}\.t
> diff \-\-git a/t/TEST b/t/TEST
> index 6124c5d\.\.b8dc4c2 100755
> \-\-\- a/t/TEST
> \+\+\+ b/t/TEST
> @&#8203;@&#8203; \-411\,7 \+411\,7 @&#8203;@&#8203; unless \(@&#8203;ARGV\) \{
>      \# then comp\, to validate that require works
>      \# then run\, to validate that \-M works
>      \# then we know we can \-MTestInit for everything else\, making life simpler
> \-    foreach my $dir \(qw\(base comp run cmd io re op uni mro\)\) \{
> \+    foreach my $dir \(qw\(base comp run cmd io lex re op uni mro\)\) \{
>      \_find\_tests\($dir\);
>      \}
>      \_find\_tests\("lib"\) unless $&#8203;::core;
> diff \-\-git a/t/lex/interpolate\.t b/t/lex/interpolate\.t
> new file mode 100644
> index 0000000\.\.3472d97
> \-\-\- /dev/null
> \+\+\+ b/t/lex/interpolate\.t
> @&#8203;@&#8203; \-0\,0 \+1\,15 @&#8203;@&#8203;
> \+BEGIN \{
> \+    require "test\.pl"
> \+\}
> \+
> \+use strict;
> \+use warnings;
> \+
> \+plan 2;
> \+
> \+\{
> \+    my %foo = \(aap => "monkey"\);
> \+    my $foo = '';
> \+    is\("@&#8203;\{\[$foo\{'aap'\}\]\}"\, 'monkey'\, 'interpolation of hash lookup with space between lexical variable and subscript'\);
> \+    is\("@&#8203;\{\[$foo \{'aap'\}\]\}"\, 'monkey'\, 'interpolation of hash lookup with space between lexical variable and subscript'\);
> \+\}
> \-\- 
> 1\.6\.5
>
